File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
import heapq

DUMMY = 10 ** 10

def solution(n, roads, sources, destination):
    road_info = [[] for _ in range(n+1)]
    
    for road1, road2 in roads:
        road_info[road1].append(road2)
        road_info[road2].append(road1)
    
    distances = [DUMMY for _ in range(n+1)]
    distances[destination] = 0
    
    pq = []
    # time, current_road
    heapq.heappush(pq, [0, destination])
    
    while pq:
        time, current_road = heapq.heappop(pq)
        for road in road_info[current_road]:
            if distances[road] != DUMMY: continue
            distances[road] = time+1
            heapq.heappush(pq, [time+1, road])
    
    
    answer = list(map(lambda x: -1 if distances[x] == DUMMY else distances[x], sources))
    return answer

File: 프로그래머스/3/132266. 부대복귀/부대복귀.kt
import java.util.*

class Solution {
    fun solution(n: Int, roads: Array<IntArray>, sources: IntArray, destination: Int): IntArray {
        val roadInfo = List(size = n+1) { mutableListOf<Int>() }
        
        for ((road1, road2) in roads) {
            roadInfo[road1].add(road2)
            roadInfo[road2].add(road1)
        }
        
        val pq = PriorityQueue<Pair<Int, Int>> { a1, b1 ->
            a1.first - b1.first
        }
        pq.add(0 to destination)
        val distances = MutableList<Int>(size = n+1) { Int.MIN_VALUE }
        distances[destination] = 0
        
        while (pq.isNotEmpty()) {
            val (time, currentRoad) = pq.poll()
            for (road in roadInfo[currentRoad]) {
                if (distances[road] == Int.MIN_VALUE) {
                    distances[road] = time+1
                    pq.add(time+1 to road)
                }
            }
        }
        
    
        return sources.map { 
            if (distances[it] == Int.MIN_VALUE) -1
            else distances[it]
        }.toIntArray()
    }
}